### PR TITLE
DCOS-14010: Add container to pick logging component and API

### PR DESCRIPTION
**Pull request by mlunoe**
---
Note: DCOS-14078 will take care of the breadcrumb bug. It is not handled in this PR!
~Dependent on this PR: https://github.com/dcos/dcos-ui/pull/1907~ (merged)

---

This PR adds functionality to choose one of the existing logging APIs. To use Mesos API set this variable in your `Config.dev.js`:
```js
mesos: {
  'logging-strategy': 'logrotate'
}
```
To use journald API set this variable in your `Config.dev.js`:
```js
mesos: {
  'logging-strategy': 'journald'
}
```

It will limit the files to view to be only `stdout` and `stderr`, when using Mesos API. Other files will still be available to view using the file viewer under the files tab.

Using Mesos API:
Mesos API
Using `journald` API
journald API

**Checklist**
- [x] Did you add a JIRA issue in a commit message or as part of the branch name?
- [x] Did you add new unit tests?
- [ ] Did you add new integration tests?
- [ ] ~~If this is a regression, did you write a test to catch this in the future?~~ (Not applicable)

~~For reviewers: View diff between PRs [here](https://github.com/dcos/dcos-ui/compare/mlunoe/bug-fix/logs-unsubscribing-on-file-load…mlunoe/feature/DCOS-14010-add-container-to-pick-logging-component-and-api?expand=1)~~ Merged!

### CODE REVIEW

**Comment by mlunoe** [comment]
Run integration tests

**Comment by jfurrow** [comment]
When I set it to `logrotate` it seems that there are some problems with going backwards in the log viewer. Also the "Go to bottom" button doesn't ever go away, but I'm not sure that's related to this PR.

Here's a screen recording: https://cl.ly/2O0R2Z3C2G16

**Comment by orlandohohmeier** [comment]
I ran into the following errors while testing this and don't feel comfortable green-lighting this one without the ability to properly test it esp. as the loading was somewhat flaky and is very slow.

- https://jira.mesosphere.com/browse/DCOS_OSS-715
- https://jira.mesosphere.com/browse/DCOS_OSS-718
- https://jira.mesosphere.com/browse/DCOS_OSS-717
- https://jira.mesosphere.com/browse/DCOS_OSS-716

**Comment by mlunoe** [comment]
FYI: Number 1. and 4. is because this backend fix has not merged yet: https://github.com/mesosphere/dcos-enterprise/pull/654

**Comment by mlunoe** [comment]
NB: I changed the base of this point to a feature branch! Also, I put up a PR for the MesosLogView fixes here: https://github.com/dcos/dcos-ui/pull/1974

I still need to fix these two issues:
- https://jira.mesosphere.com/browse/DCOS_OSS-718
- https://jira.mesosphere.com/browse/DCOS_OSS-717

And as mentioned before these have been fixed in the backend:
https://jira.mesosphere.com/browse/DCOS_OSS-715
https://jira.mesosphere.com/browse/DCOS_OSS-716

File: plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  render() {
+    if (this.state.isLoading) {
+      return <Loader />;
+    }
+
+    const {props} = this;
+    const config = ConfigStore.get('config');
+    const useSystemLogsAPI = findNestedPropertyInObject(
```
Comment (orlandohohmeier) [inline plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js]: Please don't mix conditions and statements, as those lines are rather hard to follow. 

```js
const loggingStrategy = findNestedPropertyInObject(
    config, 
    'uiConfiguration.plugins.mesos.logging-strategy'
);

 if (loggingStrategy === SYSTEM_LOGS) {
    return <TaskSystemLogsContainer {...props} />;
 }
```

File: plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      config,
+      'uiConfiguration.plugins.mesos.logging-strategy'
+    ) === SYSTEM_LOGS;
+
+    if (useSystemLogsAPI) {
+      return <TaskSystemLogsContainer {...props} />;
+    }
+
+    return <TaskFileViewer limitLogFiles={['stdout', 'stderr']} {...props} />;
```
Comment (orlandohohmeier) [inline plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js]: Do we really need to pass all the props? Passing all of the is an anti-pattern that should be avoided. Please use concrete/defined properties instead.

**Comment by MatApple** [comment]
When using the mesos API I get this warning when I scroll in the log window:

File: plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js (hunk 1 above)
Comment (rcorral) [inline plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js]: Fixed!

File: plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import {StoreMixin} from 'mesosphere-shared-reactjs';
+
+import ConfigStore from '../../../../../../src/js/stores/ConfigStore';
+import Loader from '../../../../../../src/js/components/Loader';
+import {SYSTEM_LOGS} from '../../../../../../src/js/constants/MesosLoggingStrategy';
+import TaskSystemLogsContainer from './TaskSystemLogsContainer';
+import TaskFileViewer from './TaskFileViewer';
+import {findNestedPropertyInObject} from '../../../../../../src/js/utils/Util';
+
+class TaskLogsAPIContainer extends mixin(StoreMixin) {
```
Comment (rcorral) [inline plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js]: Fixed!

File: plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js (hunk 2 above)
Comment (rcorral) [inline plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js]: looking into this now.

**Comment by MatApple** [comment]
The `Go to bottom` button seems a little flaky (sometimes shows, sometimes doesn't, sometimes shows when at bottom already) but I don't see it as a blocker.

**Comment by jfurrow** [comment]
@MatApple The scrolling `setState` error will be fixed by updating to the next release of `reactjs-components`

**Comment by MatApple** [comment]
We also loose the log tab highlighting sometimes -

**Comment by rcorral** [comment]
@MatApple that scrolling error isn't related to this task. If you simply shrink your browser to a smaller size this error will happen to any container that has a scrollbar.
This is in reference to [this](https://github.com/dcos/dcos-ui/pull/1954#issuecomment-285142853) comment.

**Comment by MatApple** [comment]
Looking through the remaining open PRs that follow this one I can see @mlunoe has fixed the bugs that are noted as outstanding on this PR (Tab highlight and Go to bottom flakiness) https://github.com/dcos/dcos-ui/pull/1974#issuecomment-285096797

**Comment by MatApple** [comment]
@orlandohohmeier - @rcorral updated this PR based on your requested changes so I'm going to merge 👍

File: plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js (hunk 2 above)
Comment (rcorral) [inline plugins/services/src/js/pages/task-details/TaskLogsAPIContainer.js]: Just put up this PR #1977 to address this comment as this PR merged early.
